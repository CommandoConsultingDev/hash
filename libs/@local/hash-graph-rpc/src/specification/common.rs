use std::future::{ready, Future};

use bytes::Bytes;
use error_stack::{Context, Report, ResultExt};
use hash_status::StatusCode;
use thiserror::Error;

use crate::harpc::{Decode, Encode, Stateful};

#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[cfg_attr(feature = "specta", derive(specta::Type))]
#[cfg_attr(feature = "specta", specta(transparent))]
pub struct Error(hash_status::Status<serde_json::Value>);

impl<C> From<Report<C>> for Error
where
    C: Context,
{
    fn from(value: Report<C>) -> Self {
        let message = value.to_string();

        let status_code = value
            .request_ref::<StatusCode>()
            .next()
            .copied()
            .unwrap_or(StatusCode::Internal);

        let report = serde_json::to_value(&value).unwrap_or_else(|_| {
            tracing::error!("unable to convert error to json");
            serde_json::Value::Array(Vec::new())
        });

        let contents = match report {
            serde_json::Value::Array(reports) => reports,
            report => vec![report],
        };

        let status = hash_status::Status::new(status_code, Some(message), contents);

        Self(status)
    }
}

#[derive(Debug, Copy, Clone, Error)]
#[error("serde error")]
pub struct SerdeError;

pub struct JsonCodec;

impl Stateful for JsonCodec {
    type State = ();

    fn state(&self) -> Self::State {
        ()
    }
}

impl<T> Encode<T> for JsonCodec
where
    T: serde::Serialize,
{
    type Error = SerdeError;

    fn encode(
        &self,
        value: T,
    ) -> impl Future<Output = error_stack::Result<Bytes, Self::Error>> + Send {
        let bytes = serde_json::to_vec(&value)
            .change_context(SerdeError)
            .map(Bytes::from);

        ready(bytes)
    }
}

impl<T> Decode<T> for JsonCodec
where
    T: serde::de::DeserializeOwned + Send,
{
    type Error = SerdeError;

    fn decode(
        &self,
        bytes: Bytes,
    ) -> impl Future<Output = error_stack::Result<T, Self::Error>> + Send {
        let value = serde_json::from_slice(&bytes).change_context(SerdeError);

        ready(value)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct JsonContext<S> {
    state: S,
}

impl<S> JsonContext<S> {
    #[must_use]
    pub const fn new(state: S) -> Self {
        Self { state }
    }
}

impl<S> Stateful for JsonContext<S>
where
    S: Send + Sync + Clone,
{
    type State = S;

    fn state(&self) -> Self::State {
        self.state.clone()
    }
}

impl<S, T> Encode<T> for JsonContext<S>
where
    S: Send + Sync + Clone,
    T: serde::Serialize,
{
    type Error = SerdeError;

    fn encode(
        &self,
        value: T,
    ) -> impl Future<Output = error_stack::Result<Bytes, Self::Error>> + Send {
        JsonCodec.encode(value)
    }
}

impl<S, T> Decode<T> for JsonContext<S>
where
    S: Send + Sync + Clone,
    T: serde::de::DeserializeOwned + Send,
{
    type Error = SerdeError;

    fn decode(
        &self,
        bytes: Bytes,
    ) -> impl Future<Output = error_stack::Result<T, Self::Error>> + Send {
        JsonCodec.decode(bytes)
    }
}

impl<S> crate::harpc::Context for JsonContext<S> where S: Clone + Send + Sync + 'static {}
