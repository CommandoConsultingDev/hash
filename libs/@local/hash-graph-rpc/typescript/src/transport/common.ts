import { Brand } from "effect";

export type TransportVersion = number & Brand.Brand<"TransportVersion">;
export const TransportVersion = Brand.refined<TransportVersion>(
  (value) => Number.isInteger(value) && value >= 0 && value <= 255,
  (value) =>
    Brand.error(`TransportVersion must be between 0 and 255, got ${value}`),
);

export type ServiceVersion = number & Brand.Brand<"ServiceVersion">;
export const ServiceVersion = Brand.refined<ServiceVersion>(
  (value) => Number.isInteger(value) && value >= 0 && value <= 255,
  (value) =>
    Brand.error(`ServiceVersion must be between 0 and 255, got ${value}`),
);

export type ServiceId = number & Brand.Brand<"ServiceId">;
export const ServiceId = Brand.refined<ServiceId>(
  (value) => Number.isSafeInteger(value) && value >= 0,
  (value) => Brand.error(`ServiceId must be a positive integer, got ${value}`),
);

export type ProcedureId = number & Brand.Brand<"ProcedureId">;
export const ProcedureId = Brand.refined<ProcedureId>(
  (value) => Number.isSafeInteger(value) && value >= 0,
  (value) =>
    Brand.error(`ProcedureId must be a positive integer, got ${value}`),
);

export type ActorId = Uint8Array & Brand.Brand<"ActorId">;
// validate that it is indeed a uuid, currently not done easily
export const ActorId = Brand.refined<ActorId>(
  (value) => value.length === 16,
  (value) =>
    Brand.error(`ActorId must be a 16 bytes UUID, got ${value.length}`),
);

export type PayloadSize = number & Brand.Brand<"Size">;
export const PayloadSize = Brand.refined<PayloadSize>(
  (value) => Number.isSafeInteger(value) && value >= 0,
  (value) => Brand.error(`Size must be a positive integer, got ${value}`),
);
