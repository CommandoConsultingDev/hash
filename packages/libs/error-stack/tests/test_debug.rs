#![cfg_attr(nightly, feature(provide_any))]
// can be considered safe, because we only check the output, which in itself does not use **any**
// unsafe code.
#![cfg(not(miri))]
#![cfg_attr(all(nightly, feature = "std"), feature(error_generic_member_access))]

mod common;

use common::snapshots::*;
#[allow(unused_imports)]
use error_stack::Report;
use insta::assert_snapshot;

fn prepare(suffix: bool) -> impl Drop {
    snapshots::prepare(suffix, true)
}

/// This is the main test, to test all different parts at once,
/// and demonstrates that the rendering algorithm works at arbitrary depth.
#[test]
fn sources_nested() {
    let _guard = prepare(true);

    let report = create_sources_nested();

    assert_snapshot!(format!("{report:?}"));
}

#[test]
fn sources_nested_alternate() {
    let _guard = prepare(true);

    let report = create_sources_nested();

    assert_snapshot!(format!("{report:#?}"));
}

#[cfg(all(
    rust_1_65,
    any(feature = "std", feature = "hooks"),
    feature = "spantrace",
    feature = "pretty-print"
))]
mod full {
    //! Why so many cfg guards?
    //! What was found during initial development of the feature was,
    //! that a complete test of all tests with snapshots on every possible feature combination
    //! was infeasible, as this would lead to *a lot* of different snapshots.
    //!
    //! Changes in snapshots (this includes adding or removing lines in the test code) results in 9
    //! different snapshots, which *all* basically test the same permutation:
    //! * Does glyph/non-glyph output look nice?
    //! * Does rendering work?
    //! * Do the builtin hooks (`Backtrace` and `SpanTrace`) work?
    //!
    //! Does any combination of those work together?
    //! Therefore most of them are redundant, this means that we can cut down on the amount of
    //! snapshots that are generated.
    //! This does *not* impact speed, but makes it easier to look through all snapshots, which means
    //! that instead of 118 new snapshots once a code line changes, one just needs to look over
    //! < 30, which is a lot more manageable.
    //!
    //! There are still some big snapshot tests, which are used evaluate all of the above.

    #[cfg(nightly)]
    use std::any::Demand;
    use std::{
        error::Error,
        fmt::{Display, Formatter},
        panic::Location,
    };

    use super::*;

    #[test]
    fn linear() {
        let _guard = prepare(false);

        let report = create_report()
            .attach_printable(PrintableA(0))
            .attach(AttachmentA)
            .attach(AttachmentB)
            .change_context(ContextA(0))
            .attach_printable(PrintableB(0))
            .attach(AttachmentB)
            .change_context(ContextB(0))
            .attach_printable("printable C");

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn linear_ext() {
        let _guard = prepare(false);

        let report = create_report()
            .attach_printable(PrintableA(0))
            .attach(AttachmentA)
            .attach(AttachmentB)
            .change_context(ContextA(0))
            .attach_printable(PrintableB(0))
            .attach(AttachmentB)
            .change_context(ContextB(0))
            .attach_printable("printable C");

        assert_snapshot!(format!("{report:#?}"));
    }

    #[derive(Debug)]
    struct ContextC;

    impl Display for ContextC {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            f.write_str("A multiline\ncontext that might have\na bit more info")
        }
    }

    impl Error for ContextC {}

    #[test]
    fn multiline_context() {
        let _guard = prepare(false);

        let report = Report::new(ContextC)
            .change_context(ContextC)
            .attach_printable(PrintableB(0))
            .attach(AttachmentB)
            .change_context(ContextB(0))
            .attach_printable("printable C");

        assert_snapshot!(format!("{report:#?}"));
    }

    #[test]
    fn multiline() {
        let _guard = prepare(false);

        let report = create_report()
            .attach_printable("A multiline\nattachment\nthat might have some\nadditional info")
            .attach_printable("A multiline\nattachment\nthat might have some\nadditional info");

        assert_snapshot!(format!("{report:#?}"));
    }

    /// Generate the `Debug` for
    ///
    /// ```text
    ///         [A] B
    ///            |
    ///         [C] A
    ///            |
    ///         [A] A
    ///      /     |    \
    ///  [P] A  [P] B  [P] B
    ///    |       |      |
    ///   Root   Root    Root
    /// ```
    ///
    /// This should demonstrate that we're able to generate with multiple groups at the same time.
    #[test]
    fn sources() {
        let _guard = prepare(false);

        let mut root1 = create_report().attach_printable(PrintableA(1));
        let root2 = create_report().attach_printable(PrintableB(2));
        let root3 = create_report().attach_printable(PrintableB(3));

        root1.extend_one(root2);
        root1.extend_one(root3);

        let report = root1
            .attach(AttachmentA(1))
            .change_context(ContextA(2))
            .attach(AttachmentB(2));

        assert_snapshot!(format!("{report:?}"));
    }

    /// Generate the `Debug` for:
    ///
    /// ```text
    ///         [A] B
    ///            |
    ///         [C] A
    ///            |
    ///         [A] A
    ///           /  \
    ///       [P] A  [A] A
    ///         |       |   \
    ///        Root  [P] B  [P] B
    ///                 |      |
    ///               Root    Root
    /// ```
    ///
    /// and should demonstrate that "transparent" groups, groups which do not have a change in
    /// context, are still handled gracefully.
    #[test]
    fn sources_transparent() {
        let _guard = prepare(false);

        let report = {
            let mut report = create_report().attach_printable(PrintableA(1));

            report.extend_one({
                let mut report = create_report().attach_printable(PrintableB(2));

                report.extend_one(create_report().attach_printable(PrintableB(3)));

                report.attach_printable(PrintableA(4))
            });

            report
                .attach(AttachmentA(1))
                .change_context(ContextA(2))
                .attach(AttachmentB(2))
        };

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn complex() {
        let _guard = prepare(false);

        let mut report = create_report().attach_printable(PrintableA(0));
        report.extend_one({
            let mut report = create_report().attach_printable(PrintableB(1));

            report.extend_one(
                create_report()
                    .attach(AttachmentB(0))
                    .attach(AttachmentA(1))
                    .attach_printable(PrintableB(1)),
            );

            report.attach(AttachmentA(2)).attach_printable("Test")
        });

        // force the generation of a tree node
        let report = report
            .change_context(ContextA(2))
            .attach_printable(PrintableA(2));

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook() {
        let _guard = prepare(false);

        let report = create_report().attach(2u32);

        Report::install_debug_hook::<u32>(|_, context| {
            context.push_body("unsigned 32bit integer");
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_context() {
        let _guard = prepare(false);

        let report = create_report().attach(2u32);

        Report::install_debug_hook::<u32>(|_, context| {
            let idx = context.increment_counter();
            context.push_body(format!("unsigned 32bit integer (No. {idx})"));
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_for_context() {
        let _guard = prepare(false);

        let report = create_report().attach(2u32);

        Report::install_debug_hook::<RootError>(|_, _| {
            // This should not be displayed as `RootError` is only used as `Context`, never as
            // attachment.
            unreachable!("A context should never be used as hook");
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_multiple() {
        let _guard = prepare(false);

        let report = create_report().attach(1u32).attach(2u64);

        Report::install_debug_hook::<u32>(|_, context| {
            context.push_body("unsigned 32bit integer");
        });
        Report::install_debug_hook::<u64>(|_, context| {
            context.push_body("unsigned 64bit integer");
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_decr() {
        let _guard = prepare(false);

        let report = create_report() //
            .attach(1u32)
            .attach(2u32)
            .attach(3u32);

        Report::install_debug_hook::<u32>(|_, context| {
            let idx = context.decrement_counter();
            context.push_body(idx.to_string());
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_incr() {
        let _guard = prepare(false);

        let report = create_report() //
            .attach(1u32)
            .attach(2u32)
            .attach(3u32);

        Report::install_debug_hook::<u32>(|_, context| {
            let idx = context.increment_counter();
            context.push_body(idx.to_string());
        });

        assert_snapshot!(format!("{report:?}"));
    }

    #[test]
    fn hook_alternate() {
        let _guard = prepare(false);

        let report = create_report().attach(2u64);

        Report::install_debug_hook::<u64>(|_, context| {
            if context.alternate() {
                context.push_appendix("Snippet");
            }

            context.push_body("Empty");
        });

        assert_snapshot!("norm", format!("{report:?}"));

        assert_snapshot!("alt", format!("{report:#?}"));
    }

    #[test]
    fn hook_location() {
        let _guard = prepare(false);

        let report = create_report();

        Report::install_debug_hook::<Location<'static>>(|_, _| {});

        assert_snapshot!(format!("{report:?}"))
    }

    #[cfg(nightly)]
    #[derive(Debug)]
    struct ContextD {
        code: usize,
        reason: &'static str,
    }

    #[cfg(nightly)]
    impl Display for ContextD {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            f.write_str("context D")
        }
    }

    #[cfg(nightly)]
    impl Error for ContextD {
        fn provide<'a>(&'a self, req: &mut Demand<'a>) {
            req.provide_ref(&self.code);
            req.provide_ref(&self.reason);
        }
    }

    #[test]
    #[cfg(nightly)]
    fn hook_provider() {
        let _guard = prepare(false);

        let report = create_report().change_context(ContextD {
            code: 420,
            reason: "Invalid User Input",
        });

        Report::install_debug_hook::<usize>(|value, context| {
            context.push_body(format!("usize: {value}"));
        });
        Report::install_debug_hook::<&'static str>(|value, context| {
            context.push_body(format!("&'static str: {value}"));
        });

        assert_snapshot!(format!("{report:?}"));
    }
}
