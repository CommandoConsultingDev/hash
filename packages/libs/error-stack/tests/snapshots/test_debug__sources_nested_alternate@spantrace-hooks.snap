---
source: tests/test_debug.rs
expression: "format!(\"{report:#?}\")"
---
context A
|- at tests/common/snapshots.rs:220:10
|- 2
|- 1
|
|-> context A
 |  |- at tests/common/snapshots.rs:214:10
 |  |- 4
 |  |- 3
 |  |
 |  |-> root error
 |      |- at tests/common/mod.rs:4:5
 |      |- span trace with 2 frames (1)
 |      |- 6
 |
 |> context A
    |- at tests/common/snapshots.rs:209:10
    |- 5
    |- 3
    |
    |-> context A
    |   |- at tests/common/snapshots.rs:207:10
    |   |- 7
    |
    |-> context A
     |  |- at tests/common/snapshots.rs:198:34
     |  |- 9
     |  |- 8
     |  |
     |  |-> root error
     |      |- at tests/common/mod.rs:4:5
     |      |- span trace with 2 frames (2)
     |
     |> context A
     |  |- at tests/common/snapshots.rs:192:10
     |  |- 13
     |  |- 10
     |  |- 16
     |  |- 9
     |  |- 8
     |  |
     |  |-> root error
     |      |- at tests/common/mod.rs:4:5
     |      |- span trace with 2 frames (3)
     |
     |> context A
     |  |- at tests/common/snapshots.rs:178:10
     |  |- 15
     |  |- 14
     |  |- 10
     |  |- 16
     |  |- 9
     |  |- 8
     |  |
     |  |-> root error
     |      |- at tests/common/mod.rs:4:5
     |      |- span trace with 2 frames (4)
     |
     |> context A
     |  |- at tests/common/snapshots.rs:188:10
     |  |- 11
     |  |- 9
     |  |- 8
     |  |
     |  |-> root error
     |      |- at tests/common/mod.rs:4:5
     |      |- span trace with 2 frames (5)
     |
     |> context A
        |- at tests/common/snapshots.rs:184:10
        |- 12
        |- 9
        |- 8
        |
        |-> context A
        |   |- at tests/common/snapshots.rs:183:10
        |
        |-> root error
            |- at tests/common/mod.rs:4:5
            |- span trace with 2 frames (6)

========================================

span trace No. 1
  [redacted]

span trace No. 2
  [redacted]

span trace No. 3
  [redacted]

span trace No. 4
  [redacted]

span trace No. 5
  [redacted]

span trace No. 6
  [redacted]
